std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
+    tree.* = Tree{
         .source = source,
-        .root_node = root_node,
+        .root_node = undefined,
+        .tokens = token_list,
+        .errors = Tree.ErrorList.init(arena),
         .arena_allocator = tree_arena,
-        .tokens = ast.Tree.TokenList.init(arena),
-        .errors = ast.Tree.ErrorList.init(arena),
     };
+    tree.root_node = try parseRoot(&tree.arena_allocator.allocator, &it, tree);
+    return tree;
+}
 
-    var tokenizer = Tokenizer.init(tree.source);
-    while (true) {
-        const token_ptr = try tree.tokens.addOne();
-        token_ptr.* = tokenizer.next();
-        if (token_ptr.id == Token.Id.Eof) break;
-    }
-    var tok_it = tree.tokens.iterator(0);
+// Root <- skip ContainerMembers eof
+fn parseRoot(arena: *Allocator, it: *TokenIterator, tree: *Tree) Allocator.Error!*Node.Root {
+    const node = try arena.create(Node.Root);
+    node.* = Node.Root{
+        .base = Node{ .id = .Root },
+        .decls = undefined,
+        // TODO: File an issue: Currently it is impossible to have a container-level
+        //       doc comment and NO doc comment on the first decl. For now, simply
+        //       ignore the problem and assume that there will be no container-level
+        //       doc comments.
+        .doc_comments = null,
+        .eof_token = undefined,
+    };
+    node.decls = parseContainerMembers(arena, it, tree, .Keyword_struct) catch |err| {
+        // TODO: Switch on the error type
+        // https://github.com/ziglang/zig/issues/2203
+        // return switch (err) {
+        //     Allocator.Error => |e| e,
+        //     Error.UnexpectedToken => tree,
+        // };
+        if (err == Error.UnexpectedToken) return node;
+        assert(err == Allocator.Error.OutOfMemory);
+        return Allocator.Error.OutOfMemory;
+    };
+    node.eof_token = eatToken(it, .Eof) orelse {
+        try tree.errors.push(AstError{
+            .ExpectedContainerMembers = AstError.ExpectedContainerMembers{ .token = it.index },
+        });
+        return node;
+    };
+    return node;
+}
+
+// ContainerMembers
+//     <- TestDecl ContainerMembers
+//      / TopLevelComptime ContainerMembers
+//      / KEYWORD_pub? TopLevelDecl ContainerMembers
+//      / KEYWORD_pub? ContainerField COMMA ContainerMembers
+//      / KEYWORD_pub? ContainerField
+//      /
+fn parseContainerMembers(arena: *Allocator, it: *TokenIterator, tree: *Tree, kind: Token.Id) !Node.Root.DeclList {
+    var list = Node.Root.DeclList.init(arena);
 
-    // skip over line comments at the top of the file
     while (true) {
-        const next_tok = tok_it.peek() orelse break;
-        if (next_tok.id != Token.Id.LineComment) break;
-        _ = tok_it.next();
-    }
+        const doc_comments = try parseDocComment(arena, it, tree);
 
-    try stack.append(State.TopLevel);
+        if (try parseTestDecl(arena, it, tree)) |node| {
+            node.cast(Node.TestDecl).?.doc_comments = doc_comments;
```

Comment:
Why not avoid this cast by having `parseTestDecl` return `*Node.TestDecl`? I see this pattern in a few other places.

Reply:
I agree it would be nice to avoid the casts. We could probably do this for a bunch of these rule parsers. However I think this might conflict (philosophically) with what @Hejsil requested [over here](https://github.com/ziglang/zig/pull/2405#discussion_r282834107) when one of the rule parsers deviated from the common signature: 

> We should strive to have all functions follow the ParseFn signature when possible

I might be reading too far into it, of course. Anyway, if there's any conflict of opinion I think I'll leave it as-is for now and revisit another time, if that's cool!

Reply:
(Of course, I should note that by necessity, all of the rule parsers that return a `Node.SomeKindaList` type already deviate from this signature by not returning a `*Node`. FWIW.)